AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +615,840 @@ class AzureIoTHub {
-        function _resetConnectionTables() {
-            _sessions = _getDefaultConnectionTable();
-            _transfers = _getDefaultConnectionTable();
-            _senders = _getDefaultConnectionTable();
-            _receivers = _getDefaultConnectionTable();
-            _msgQueue = [];
-            _handlers = {};
+        function _handleCloudToDevMsg(message, topic) {
+            local props = {};
+            local splittedTopic = split(topic, "/");
+            if (splittedTopic[splittedTopic.len() - 1] != "devicebound") {
+                // We have properties sent with message
+                props = http.urldecode(splittedTopic[splittedTopic.len() - 1]);
+            }
+            _onMessageCb(AzureIoTHub.Message(message, props));
+        }
+
+        function _handleDesPropsMsg(message, topic) {
+            local version = null;
+            local parsedMsg = null;
+            try {
+                version = split(topic, "=")[1];
+                parsedMsg = http.jsondecode(message);
+            } catch (e) {
+                _logError("Exception at parsing the message: " + e);
+                _logMsg(message, topic);
+                return;
+            }
+            _onTwinReqCb(version, parsedMsg);
+        }
+
+        function _handleTwinResponse(message, topic) {
+            local status = null;
+            local reqId = null;
+            local parsedMsg = null;
+            try {
+                local splitted = split(topic, "/");
+                status = splitted[3].tointeger();
+                reqId = http.urldecode(splitted[4])["?$rid"];
+                if (_statusIsOk(status) && message != null && message != "") {
+                    parsedMsg = http.jsondecode(message);
+                }
+            } catch (e) {
+                _logError("Exception at parsing the message: " + e);
+                _logMsg(message, topic);
+                return;
+            }
+            // Twin's properties received after UPDATE request
+            if (reqId in _twinPendingRequests) {
+                _handleTwinUpdateResponse(reqId, status);
+            // Twin's properties received after GET request
+            } else if (_twinRetrievedCb != null && _twinRetrievedCb[0] == reqId) {
+                _handleTwinGetResponse(reqId, status, parsedMsg);
+            } else {
+                _log("Message with unknown request ID received: ");
+                _logMsg(message, topic);
+            }
         }
 
-        function _getDefaultConnectionTable() {
-            return {"event" : null, "auth" : null};
+        function _handleTwinUpdateResponse(reqId, status) {
+            local arr = delete _twinPendingRequests[reqId];
+            // If no pending requests, cancel the timer
+            if (_twinRetrievedCb == null && _twinPendingRequests.len() == 0) {
+                imp.cancelwakeup(_processQueueTimer);
+                _processQueueTimer = null;
+            }
+            local props = arr[0];
```

Comment:
I would make these indices constants to make it easier to read. i.e. TWIN_PROPERTIES_INDEX, etc. You can define constants within the scope of the function or in the constructor (if you want to reuse them in multiple methods like in this case).

Constants cost nothing in terms of footprint as they are propagated at a special preprocessor step during compilation.

Reply:
Done